mllib-local/src/main/scala/org/apache/spark/ml/linalg/Vectors.scala

```diff
@@ -603,6 +603,19 @@ class SparseVector @Since("2.0.0") (
 
   private[spark] override def asBreeze: BV[Double] = new BSV[Double](indices, values, size)
 
+  override def apply(i: Int): Double = {
+    if (i < 0 || i >= size) {
+      throw new IndexOutOfBoundsException(s"Index $i out of bounds [0, $size)")
+    }
+
+    if (indices.isEmpty || i < indices(0) || i > indices(indices.length - 1)) {
```

Comment:
Isn't this case already covered by the binarySearch case below? if it's not found for any reason you get a negative results.

Reply:
1, the impl of `Arrays.binarySearch` does not chek the range:
```
    public static int binarySearch(int[] a, int key) {
        return binarySearch0(a, 0, a.length, key);
    }

    // Like public version, but without range checks.
    private static int binarySearch0(long[] a, int fromIndex, int toIndex,
                                     long key) {
        int low = fromIndex;
        int high = toIndex - 1;

        while (low <= high) {
            int mid = (low + high) >>> 1;
            long midVal = a[mid];

            if (midVal < key)
                low = mid + 1;
            else if (midVal > key)
                high = mid - 1;
            else
                return mid; // key found
        }
        return -(low + 1);  // key not found.
    }
```
2, in `breeze.collection.mutable.SparseArray`, the `findOffset` function called in `apply` to perform binary seach, take the special case that the key is out of range into account
```
    if (used == 0) {
      // empty list do nothing
      -1
    } else {
      val index = this.index
      if (i > index(used - 1)) {
        // special case for end of list - this is a big win for growing sparse arrays
        ~used
```

so I added those simple checking between binary search.

Reply:
Yes you need the check that the index is < 0 or >= length, keep that.
But `binarySearch` already handles the case that the query index is >= 0 but before the first actual index:
```
scala> java.util.Arrays.binarySearch(Array(2,3), 1)
res0: Int = -1

scala> java.util.Arrays.binarySearch(Array(2,3), 4)
res1: Int = -3
```
Why repeat that part?

Reply:
I see. This performance improvement comes from avoiding to walk a binary tree if a `key` is not included in a given sorted array. It makes sense to me .

One question. What do you mean `avoiding internal conversion` in the description?

Reply:
Yeah, but you're also always paying the cost of these two checks. It depends on the access pattern, but assuming pretty uniform distribution, the check will rarely save checks and always add a few. It seems simpler to avoid it unless there's a clear case it's a win.

Reply:
@srowen @kiszk 
on each call of `Sparse.apply`, a conversion to `breeze.linalg.SparseVector` & `breeze.collection.mutable.SparseArray` was performed internally.
The improvement coms from avoiding this.

Reply:
you can see that if the `nnz` grows, the speed up decrese. That is because with a big `nnz`, the  searching complexity `log(nnz)` dominate the whole process. However, when `nnz` is a small number (most frequently), the conversion is relatively the main part.

Reply:
@srowen I add the checks just because in the impl of `findOffset` in  `breeze.collection.mutable.SparseArray`,
it says `// special case for end of list - this is a big win for growing sparse arrays`, and I think it is reasonable.

Reply:
@zhengruifeng Would it be possible to show the performance comparison in the case that @srowen mentions. In other words, most of keys exist in indices. I hope the overhead of adding three tests would be negligible.

Reply:
Where does a conversion happen? this is just avoiding `binarySearch`, no?

Reply:
Existing `SparseVector` do not override the `apply` method inheriting from `Vector`:
```
  /**
   * Gets the value of the ith element.
   * @param i index
   */
  @Since("2.0.0")
  def apply(i: Int): Double = asBreeze(i)
```

So a `spark.ml.linalg.SparseVector` will first be converted to a `breeze.collection.mutable.SparseArray` and then a `breeze.linalg.SparseVector`.

As to the range check, I think it is just a tiny optimization.

Reply:
Oh right I see. That's a big win.
Well I'm OK with it though still not clear the extra range checks are an optimization.